tests/lib/linter/linter.js

```diff
@@ -349,4451 +284,3751 @@ describe("Linter", () => {
-
-            linter.verify("0", config, filename);
-        });
-
-    });
-
-    describe("when a custom parser is defined using defineParser", () => {
-
-        it("should be able to define a custom parser", () => {
-            const parser = {
-                parseForESLint: function parse(code, options) {
-                    return {
-                        ast: esprima.parse(code, options),
-                        services: {
-                            test: {
-                                getMessage() {
-                                    return "Hi!";
-                                }
-                            }
-                        }
-                    };
-                }
-            };
-
-            linter.defineParser("test-parser", parser);
-            const config = { rules: {}, parser: "test-parser" };
-            const messages = linter.verify("0", config, filename);
-            const suppressedMessages = linter.getSuppressedMessages();
-
-            assert.strictEqual(messages.length, 0);
             assert.strictEqual(suppressedMessages.length, 0);
         });
-
     });
 
-    describe("when config has parser", () => {
-
-        it("should pass parser as parserPath to all rules when provided on config", () => {
-
-            const alternateParser = "esprima";
+    describe("at any time", () => {
+        const code = "filename-rule";
 
-            linter.defineParser("esprima", esprima);
-            linter.defineRule("test-rule", {
-                create: sinon.mock().withArgs(
-                    sinon.match({ parserPath: alternateParser })
-                ).returns({})
+        it("has access to the filename", () => {
+            linter.defineRule(code, {
+                create: context => ({
+                    Literal(node) {
+                        assert.strictEqual(context.getFilename(), context.filename);
+                        context.report(node, context.filename);
+                    }
+                })
             });
 
-            const config = { rules: { "test-rule": 2 }, parser: alternateParser };
-
-            linter.verify("0", config, filename);
-        });
+            const config = { rules: {} };
 
-        it("should use parseForESLint() in custom parser when custom parser is specified", () => {
-            const config = { rules: {}, parser: "enhanced-parser" };
+            config.rules[code] = 1;
 
-            linter.defineParser("enhanced-parser", testParsers.enhancedParser);
             const messages = linter.verify("0", config, filename);
             const suppressedMessages = linter.getSuppressedMessages();
 
-            assert.strictEqual(messages.length, 0);
+            assert.strictEqual(messages[0].message, filename);
             assert.strictEqual(suppressedMessages.length, 0);
         });
 
-        it("should expose parser services when using parseForESLint() and services are specified", () => {
```

Comment:
I think we should keep tests with `parserServices` as they're testing how Linter creates SourceCode objects based on parser features. We could just remove the `assert.strictEqual(context.parserServices, context.sourceCode.parserServices)` line.